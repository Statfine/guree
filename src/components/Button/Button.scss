@import "../../common/utils/index.scss";

.#{scopedClass(button)} {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background-color: #fff;
  border: 1px solid currentColor;
  border-radius: 0.3em;
  cursor: pointer;

  & + & {
    margin-left: 8px;
  }
  &.large {
    padding: 10px 30px;
  }
  &.middle {
    padding: 6px 20px;
  }
  &.small {
    padding: 2px 10px;
  }
  > .action + span {
    margin-left: 6px;
  }
  &.default {
    color: #666;
    background-color: #fff;
  }
  &.primary {
    color: #fff;
    background-color: #409eff;
  }
  &.danger {
    color: #fff;
    background-color: #f56c6c;
  }
  &.warning {
    color: #fff;
    background-color: #e6a23c;
  }
  &.text {
    color: #fff;
    background-color: #67c23a;
  }
  &[disabled] {
    opacity: 0.35;
    cursor: default;
  }

  &:focus{
    outline: none;
    opacity: 0.8;
  }
}
