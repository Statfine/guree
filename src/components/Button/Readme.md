### 基本用法

```jsx
<Button>button</Button>
```

### 按钮大小

```jsx
<Button size="large">large</Button>
<Button size="middle">middle</Button>
<Button size="small">small</Button>
```

### 设置图标

```jsx
<Button icon="adn">button</Button>
```

### 加载状态

```jsx
<Button loading>loading</Button>
```

### 禁用状态

```jsx
<Button disabled>disabled</Button>
```

### 按钮类型

```jsx
<Button type="default">default</Button>
<Button type="primary">primary</Button>
<Button type="danger">danger</Button>
<Button type="warning">warning</Button>
<Button type="text">text</Button>
```
