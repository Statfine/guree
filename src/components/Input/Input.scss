@import "../../common/utils/index.scss";

.#{scopedClass(input)} {
  display: flex;
  border: 1px solid #ddd;
  color: #aaa;
  &-affix-wrapper {
    display: flex;
    flex: 1;
    padding: 4px 10px;
    .#{scopedClass(input-content)} {
      padding: 0;
    }
  }
  &-affix {
    flex: 0;
    @include flex-center-center();
  }
  &-pre-fix {
    margin-right: 4px;
  }
  &-content {
    flex: 1;
    line-height: 1.5em;
    padding: 4px 10px;
    border: none;
    outline: none;
    cursor: inherit;
    background-color: inherit;
    color: inherit;
  }
  &-suf-fix {
    margin-left: 4px;
  }
  &-add-on {
    @include flex-center-center();
    padding: 0 10px;
    flex: 0;
    color: #999;
    background-color: #eee;
  }
  &-add-on-before {
    border-right: 1px solid #ddd;
  }
  &-add-on-after {
    border-left: 1px solid #ddd;
  }
  &-clear{
    margin-right: 4px;
  }
  &-disabled{
    cursor: not-allowed;
    background-color: #f5f5f5;
    color: rgba($color: #000000, $alpha: .25);
  }
}
