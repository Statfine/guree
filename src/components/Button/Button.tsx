import React from "react";

import "./Button.scss";

/**
 * The only true button.
 */

export interface Props {
  children: string;
}

export default function Button({ children }: Props) {
  return <button className="button">{children}</button>;
}
