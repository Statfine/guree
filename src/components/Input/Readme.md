### 基本用法

```jsx
<Input placeholder="please input"></Input>
```

### 实时获取输入框的值

```jsx
const onChangeHandler = (e) => {
  const vlaue = e.target.value;
  console.log(vlaue);
};
<Input onChange={onChangeHandler}></Input>;
```

### 设置输入框的值

通过 value 字段改变 Input 组件的 value 时，不会触发 onChange

```jsx
import { useState } from "react";
import Button from "@components/Button";

const [value, setValue] = useState("set value");
const onChangeHandler = (e) => {
  console.log("value changed");
};
const changeValue = () => {
  setValue("change value");
};
<div>
  <Input value={value} onChange={onChangeHandler}></Input>
  <br />
  <Button onClick={changeValue}>change value</Button>
</div>;
```

### 设置前/后标签

```jsx
import Icon from "@components/Icon";

<div>
  <Input addonBefore="http://" addonAfter=".com"></Input>
  <br />
  <Input addonAfter={<Icon name="atom" fill="currentColor"></Icon>}></Input>
</div>;
```

### 设置前/后缀

```jsx
import Icon from "@components/Icon";

<div>
  <Input prefix="￥" suffix="RMB"></Input>
  <br />
  <Input
    prefix={<Icon name="user" fill="currentColor"></Icon>}
    suffix={<Icon name="exclamation-circle" fill="currentColor"></Icon>}
  ></Input>
</div>;
```

### 清空输入框

```jsx
<Input allowClear></Input>
```

### 禁用输入框

```jsx
<Input disabled></Input>
```
