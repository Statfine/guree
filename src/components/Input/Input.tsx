import React, { ReactNode, ChangeEventHandler, useState } from "react";
import "./Input.scss";
import {
  combineClasses,
  scopedClass,
  classesObj,
  updateStateOnPropChange,
  inlineFlexSpan,
} from "@utils/index";
import Icon from "@components/Icon";
import { isUndefined } from "util";

interface Props {
  /** 后遮罩 */
  addonAfter?: ReactNode;
  /** 前遮罩 */
  addonBefore?: ReactNode;
  /** 占位字符 */
  placeholder?: string;
  /** 禁用状态 */
  disabled?: boolean;
  /** 前缀 */
  prefix?: ReactNode;
  /** 后缀 */
  suffix?: ReactNode;
  /** 类型，同 input 原生标签一致，除了 textarea */
  type?: string;
  /** 输入值 */
  value?: string;
  /** 输入值变化时的回调 */
  onChange?: (value: string) => {};
  /** 显示清空按钮 */
  allowClear?: boolean;
  /** 点击清空按钮的回调 */
  onClickClear?: () => {};
  /** 默认值 */
  defaultValue?: string;
}

const Input: React.FunctionComponent<Props> = ({
  addonAfter,
  addonBefore,
  placeholder,
  disabled,
  prefix,
  suffix,
  type,
  value,
  onChange,
  allowClear,
  onClickClear,
  defaultValue,
}: Props) => {
  const InputClass = (...classes: (string | Array<string> | classesObj)[]) =>
    scopedClass("input", ...classes);

  const InputClassNames = (disabled?: boolean) => {
    return combineClasses(InputClass(), InputClass(disabled ? "disabled" : ""));
  };

  const [realValue, setValue] = useState(
    isUndefined(value) ? defaultValue : value
  );

  updateStateOnPropChange(value, setValue);

  const changeHandler: ChangeEventHandler<HTMLInputElement> = (e) => {
    const { value: inputValue } = e.target;
    onChange && onChange(inputValue || "");
    if (disabled || !isUndefined(value)) {
      return;
    }
    setValue(inputValue || "");
  };

  const clear = () => {
    onClickClear && onClickClear();
    if (disabled || !isUndefined(value)) {
      return;
    }
    setValue("");
  };

  const PureInput = (
    <input
      className={InputClass("content")}
      type={type === "textarea" ? undefined : type}
      placeholder={placeholder}
      value={realValue || ""}
      onChange={changeHandler}
      disabled={disabled}
    />
  );

  return (
    <div className={InputClassNames(disabled)}>
      {addonBefore ? (
        <div
          className={combineClasses(
            InputClass("add-on"),
            InputClass("add-on-before")
          )}
        >
          {addonBefore}
        </div>
      ) : null}
      {prefix || suffix || allowClear ? (
        <div className={InputClass("affix-wrapper")}>
          <div
            className={combineClasses(
              InputClass("affix"),
              InputClass("pre-fix")
            )}
          >
            {prefix}
          </div>
          {PureInput}
          <div
            className={combineClasses(
              InputClass("affix"),
              InputClass("suf-fix")
            )}
          >
            {allowClear && realValue ? (
              <span
                className={combineClasses(inlineFlexSpan, InputClass("clear"))}
              >
                <Icon
                  name="times-circle"
                  fill="currentColor"
                  onClick={clear}
                ></Icon>
              </span>
            ) : null}
            {suffix}
          </div>
        </div>
      ) : (
        PureInput
      )}
      {addonAfter ? (
        <div
          className={combineClasses(
            InputClass("add-on"),
            InputClass("add-on-after")
          )}
        >
          {addonAfter}
        </div>
      ) : null}
    </div>
  );
};

export default Input;
