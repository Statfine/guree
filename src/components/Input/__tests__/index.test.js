import React, { useState } from "react";
import renderer from "react-test-renderer";
import Input from "..";
import { mount, shallow } from "enzyme";
import Button from "@components/Button";

describe("Input 组件", () => {
  it("正确导出 Input 组件", () => {
    const input = renderer.create(<Input></Input>).toJSON();
    expect(input).toMatchSnapshot();
  });

  it("测试 Input 组件 addonAfter 参数", () => {
    const input = renderer.create(<Input addonAfter=".com"></Input>).toJSON();
    expect(input).toMatchSnapshot();
  });

  it("测试 Input 组件 addonBefore 参数", () => {
    const input = renderer
      .create(<Input addonBefore="http://"></Input>)
      .toJSON();
    expect(input).toMatchSnapshot();
  });

  it("测试 Input 组件 placeholder 参数", () => {
    const input = renderer
      .create(<Input placeholder="placeholder"></Input>)
      .toJSON();
    expect(input).toMatchSnapshot();
  });

  it("测试 Input 组件 disabled 参数", () => {
    const input = renderer.create(<Input disabled></Input>).toJSON();
    expect(input).toMatchSnapshot();
  });

  it("测试 Input 组件 prefix 参数", () => {
    const input = renderer.create(<Input prefix="￥"></Input>).toJSON();
    expect(input).toMatchSnapshot();
  });

  it("测试 Input 组件 suffix 参数", () => {
    const input = renderer.create(<Input suffix="RMB"></Input>).toJSON();
    expect(input).toMatchSnapshot();
  });

  it("测试 Input 组件 type 参数", () => {
    const input1 = renderer.create(<Input type="number"></Input>).toJSON();
    const input2 = renderer.create(<Input type="textarea"></Input>).toJSON();
    expect(input1).toMatchSnapshot();
    expect(input2).toMatchSnapshot();
  });

  it("测试 Input 组件 value 参数", () => {
      const input = renderer.create(<Input value="value"></Input>).toJSON();
      expect(input).toMatchSnapshot();
  });

  it("测试 Input 组件 value 参数改变", () => {
      const input = mount(<Input value=""></Input>)
      input.setProps({value: "value Changed"})
      input.update()
      expect(input.find("input").prop("value")).toEqual("value Changed")
  });

  it("测试 Input 组件 allowClear 参数", () => {
    const input1 = renderer.create(<Input allowClear></Input>).toJSON();
    const input2 = renderer
      .create(<Input value="value" allowClear></Input>)
      .toJSON();
    expect(input1).toMatchSnapshot();
    expect(input2).toMatchSnapshot();
    mount(<Input value="value" allowClear></Input>)
      .find(".guree-input-clear svg")
      .simulate("click");
  });

  it("测试 Input 组件 onChange 参数", () => {
    const fn = jest.fn();
    const input = mount(<Input onChange={fn}></Input>);
    input.find("input").simulate("change", { target: { value: "1" } });
    input.find("input").simulate("change", { target: { value: undefined } });
    expect(fn).toBeCalled();
  });
});
