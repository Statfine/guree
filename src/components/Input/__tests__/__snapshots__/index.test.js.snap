// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input 组件 正确导出 Input 组件 1`] = `
<div
  className="guree-input guree-input"
>
  <input
    className="guree-input-content"
    onChange={[Function]}
    value=""
  />
</div>
`;

exports[`Input 组件 测试 Input 组件 addonAfter 参数 1`] = `
<div
  className="guree-input guree-input"
>
  <input
    className="guree-input-content"
    onChange={[Function]}
    value=""
  />
  <div
    className="guree-input-add-on guree-input-add-on-after"
  >
    .com
  </div>
</div>
`;

exports[`Input 组件 测试 Input 组件 addonBefore 参数 1`] = `
<div
  className="guree-input guree-input"
>
  <div
    className="guree-input-add-on guree-input-add-on-before"
  >
    http://
  </div>
  <input
    className="guree-input-content"
    onChange={[Function]}
    value=""
  />
</div>
`;

exports[`Input 组件 测试 Input 组件 allowClear 参数 1`] = `
<div
  className="guree-input guree-input"
>
  <div
    className="guree-input-affix-wrapper"
  >
    <div
      className="guree-input-affix guree-input-pre-fix"
    />
    <input
      className="guree-input-content"
      onChange={[Function]}
      value=""
    />
    <div
      className="guree-input-affix guree-input-suf-fix"
    />
  </div>
</div>
`;

exports[`Input 组件 测试 Input 组件 allowClear 参数 2`] = `
<div
  className="guree-input guree-input"
>
  <div
    className="guree-input-affix-wrapper"
  >
    <div
      className="guree-input-affix guree-input-pre-fix"
    />
    <input
      className="guree-input-content"
      onChange={[Function]}
      value="value"
    />
    <div
      className="guree-input-affix guree-input-suf-fix"
    >
      <span
        className="guree-span guree-input-clear"
      >
        <svg
          className="guree-icon"
          onClick={[Function]}
          style={
            Object {
              "fill": "currentColor",
              "height": "1em",
              "width": "1em",
            }
          }
        >
          <use
            xlinkHref="#times-circle"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`Input 组件 测试 Input 组件 disabled 参数 1`] = `
<div
  className="guree-input guree-input-disabled"
>
  <input
    className="guree-input-content"
    disabled={true}
    onChange={[Function]}
    value=""
  />
</div>
`;

exports[`Input 组件 测试 Input 组件 placeholder 参数 1`] = `
<div
  className="guree-input guree-input"
>
  <input
    className="guree-input-content"
    onChange={[Function]}
    placeholder="placeholder"
    value=""
  />
</div>
`;

exports[`Input 组件 测试 Input 组件 prefix 参数 1`] = `
<div
  className="guree-input guree-input"
>
  <div
    className="guree-input-affix-wrapper"
  >
    <div
      className="guree-input-affix guree-input-pre-fix"
    >
      ￥
    </div>
    <input
      className="guree-input-content"
      onChange={[Function]}
      value=""
    />
    <div
      className="guree-input-affix guree-input-suf-fix"
    />
  </div>
</div>
`;

exports[`Input 组件 测试 Input 组件 suffix 参数 1`] = `
<div
  className="guree-input guree-input"
>
  <div
    className="guree-input-affix-wrapper"
  >
    <div
      className="guree-input-affix guree-input-pre-fix"
    />
    <input
      className="guree-input-content"
      onChange={[Function]}
      value=""
    />
    <div
      className="guree-input-affix guree-input-suf-fix"
    >
      RMB
    </div>
  </div>
</div>
`;

exports[`Input 组件 测试 Input 组件 type 参数 1`] = `
<div
  className="guree-input guree-input"
>
  <input
    className="guree-input-content"
    onChange={[Function]}
    type="number"
    value=""
  />
</div>
`;

exports[`Input 组件 测试 Input 组件 type 参数 2`] = `
<div
  className="guree-input guree-input"
>
  <input
    className="guree-input-content"
    onChange={[Function]}
    value=""
  />
</div>
`;

exports[`Input 组件 测试 Input 组件 value 参数 1`] = `
<div
  className="guree-input guree-input"
>
  <input
    className="guree-input-content"
    onChange={[Function]}
    value="value"
  />
</div>
`;
